Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.m

```diff
@@ -0,0 +1,311 @@
+#import "Crashlytics/Crashlytics/Public/FirebaseCrashlytics/FIRCrashlyticsReport.h"
+
+@interface FIRCLSMetricKitManager () {
+  NSLock *_mutex;
+  FBLPromise *_metricKitDataAvailable;
+  FIRCLSExistingReportManager *_existingReportManager;
+  FIRCLSFileManager *_fileManager;
+  FIRCLSManagerData *_managerData;
+}
+
+@end
+
+@implementation FIRCLSMetricKitManager
+
+- (instancetype)initWithManagerData:(FIRCLSManagerData *)managerData
+              existingReportManager:(FIRCLSExistingReportManager *)existingReportManager
+                        fileManager:(FIRCLSFileManager *)fileManager {
+  _existingReportManager = existingReportManager;
+  _fileManager = fileManager;
+  _managerData = managerData;
+  return self;
+}
+
+/*
+ * Registers the MetricKit manager to receive MetricKit reports by adding self to the
+ * MXMetricManager subscribers. Also initializes the promise that we'll use to ensure that any
+ * MetricKit report files are included in Crashylytics fatal reports. If no crash occurred on the
+ * last run of the app, this promise is immediately resolved so that the upload of any nonfatal
+ * events can proceed.
+ */
+- (void)registerMetricKitManager {
+  [[MXMetricManager sharedManager] addSubscriber:self];
+  _metricKitDataAvailable = [FBLPromise pendingPromise];
+
+  // If there was no crash on the last run of the app, then we aren't expecting a MetricKit
+  // diagnostic report and should resolve the promise immediately.
+  NSString *crashedMarkerFileName = [NSString stringWithUTF8String:FIRCLSCrashedMarkerFileName];
+  NSString *crashedMarkerFileFullPath =
+      [[_fileManager rootPath] stringByAppendingPathComponent:crashedMarkerFileName];
+  if (![_fileManager fileExistsAtPath:crashedMarkerFileFullPath]) {
+    [_mutex lock];
+    [_metricKitDataAvailable fulfill:nil];
+    [_mutex unlock];
+  }
+}
+
+/*
+ * This method receives diagnostic payloads from MetricKit whenever a fatal or nonfatal MetricKit
+ * event occurs. If a fatal event, this method will be called when the app restarts. Since we're
+ * including a MetricKit report file in the Crashlytics report to be sent to the backend, we need
+ * to make sure that we process the payloads and write the included information to file before
+ * the report is sent up. If this method is called due to a nonfatal event, it will be called
+ * immediately after the event. Since we send nonfatal events on the next run of the app, we can
+ * write out the information but won't need to resolve the promise.
+ */
+- (void)didReceiveDiagnosticPayloads:(NSArray<MXDiagnosticPayload *> *)payloads {
+  NSUInteger payloadCount = [payloads count];
+  FIRCLSDebugLog(@"[Crashlytics:Crash] received %lu diagnostic payloads from MetricKit",
+                 payloadCount);
+
+  if ([payloads count] == 0) {
+    return;
+  }
+
+  MXDiagnosticPayload *diagnosticPayload = [payloads objectAtIndex:0];
+  if (!diagnosticPayload) {
+    return;
+  }
```

Comment:
Might want to iterate here? Maybe make a new method and put the rest of this method in it, with a single payload as an arg.

Reply:
My worry with that is that each type of payload has different metadata fields, so we'd end up with a lot of the same code in that new method anyway

Reply:
But is it possible that this `payloads` array has more than one element? Even if each element has multiple types of payloads themselves?

Reply:
We humaned - gonna iterate, but only take the first or last crash event